flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/PartitionTransformUdf.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the Licenet ideajoinet ideajoin for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.sink;
+
+import org.apache.flink.table.annotation.DataTypeHint;
+import org.apache.flink.table.annotation.InputGroup;
+import org.apache.flink.table.functions.ScalarFunction;
+import org.apache.iceberg.expressions.Literals;
+import org.apache.iceberg.transforms.Transform;
+import org.apache.iceberg.transforms.Transforms;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
+
+public class PartitionTransformUdf {
+
+  public static class Truncate extends ScalarFunction {
+    public String eval(int num, @DataTypeHint(inputGroup = InputGroup.ANY) Object obj) {
```

Comment:
Is there any way for us to be more specific about what types this function can take? Does Flink use these annotations to check the users SQL before running the query?

Reply:
Because I used `InputGroup.ANY`, in terms of parameter types, flink will pass any data type and finally rely on the original process to check the data type. The specific check is in the `get` function of the corresponding transform, for example
``` java
  static <T> Truncate<T> get(Type type, int width) {
    Preconditions.checkArgument(width > 0,
        "Invalid truncate width: %s (must be > 0)", width);

    switch (type.typeId()) {
      case INTEGER:
        return (Truncate<T>) new TruncateInteger(width);
      case LONG:
         .......
      default:
        throw new UnsupportedOperationException(
            "Cannot truncate type: " + type);
    }
  }
```

Both the bucket function and the truncate function can support many parameter types. If they are all split into functions with specific parameter types, it will be redundant, because only the input parameter types are different, and the function processing process is exactly the same.
Of course, we can know which types these functions support through the case branch above. This may not seem very straightforward...

Reply:
Yeah i don't see any point in adding that redundancy. If the function throws, will we get `null` or will the user's program terminate?

I know there's been some debate about this with other functions. I believe it was the introduction of `try_cast` which uses old `cast` semantics, though I don't recall the details.

Reply:
Follow up question: can we define our own `InputGroup`? And if we did, would the SQL compiler fail to build the program or would it still just fail at runtime?

I guess if the annotation is tightened, what will flink do when it goes to compile a program out of SQL that passes the wrong type to a `DataHint`?

Reply:
> can we define our own InputGroup

We can customize input and output by overriding `getTypeInference` method:
```java
@Override
public TypeInference getTypeInference(DataTypeFactory typeFactory) {
    return TypeInference.newBuilder()
            .inputTypeStrategy(...)       // custom input
            .outputTypeStrategy(...)      // custom outpout
            .build();
}
```

> And if we did, would the SQL compiler fail to build the program or would it still just fail at runtime?

If you throw an exception in `inferInputTypes`, then flink will throw exception and terminate the program:
```java
@Override
public TypeInference getTypeInference(DataTypeFactory typeFactory) {
    return TypeInference.newBuilder()
        .inputTypeStrategy(new InputTypeStrategy() {
        ...
        @Override
        public Optional<List<DataType>> inferInputTypes(CallContext callContext, boolean throwOnFailure) {
            throw new UnsupportedOperationException("exception");
        }
        ...
        })
        .build();
}

org.apache.flink.table.api.ValidationException: SQL validation failed.
...
Caused by: java.lang.UnsupportedOperationException: exception
```




> I guess if the annotation is tightened, what will flink do when it goes to compile a program out of SQL that passes the wrong type to a DataHint?

Yes, if parameter type is not matched `eval`'s parameter type , flink will throw an exception directly:
```
org.apache.flink.table.api.ValidationException: SQL validation failed. Invalid function call
```

Reply:
It seems like maybe adding the TypeInference might be a good way to go.

In addition to adding a specific function for each type.

That way, if I'm not mistaken, the compiler will be able to reject any invalid input. Then, as Ryan mentioned, we should have a function for each possible type and the correct function should be picked ahead of time based on type from the compiler.

This would avoid the runtime reflection used on each function call which has large overheard and should be avoided whenever possible.

Reply:
Now it has been changed to the way that each parameter corresponds to a function implementation.